File: javascript/5-On2_naive_pairs.js
/* 
    On2_naive_pairs.js
    Big-O: O(n²) – quadratic
    Compare every pair of elements (e.g., brute-force duplicate search). 
*/

import { performance } from "node:perf_hooks";

const hasDuplicate = (arr) => {
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[i] === arr[j]) return true;
    }
  }
  return false;
};

const main = () => {
  const N = 30000; // 30k² ≈ 900M comparisons ⇒ slow
  const nums = Array.from({ length: N }, (_, i) => i);
  nums.push(N - 1); // guarantee duplicate at the tail (worst-case)
  const t0 = performance.now();
  const dup = hasDuplicate(nums);
  const t1 = performance.now();
  console.log("duplicate found:", dup);
  console.log(`O(n²) demo with N=${N} took ${(t1 - t0).toFixed(3)} ms`);
};

main();

File: golang/5-on2/5-On2_naive_pairs.go
/*
   On2_naive_pairs.go
   Big-O: O(n²) – quadratic
   Compare every pair of elements (e.g., brute-force duplicate search).
*/

package onlogn

import (
	"fmt"
	"time"
)

func hasDuplicate(arr []int) bool {
	for i := 0; i < len(arr); i++ {
		for j := i + 1; j < len(arr); j++ {
			if arr[i] == arr[j] {
				return true
			}
		}
	}
	return false
}

func Demo() {
	N := 30000 // 30k² ≈ 900M comparisons ⇒ slow
	nums := make([]int, N)
	for i := 0; i < N; i++ {
		nums[i] = i
	}
	nums = append(nums, N-1) // guarantee duplicate at the tail (worst-case)

	t0 := time.Now()
	dup := hasDuplicate(nums)
	t1 := time.Now()

	fmt.Printf("duplicate found: %t\n", dup)
	fmt.Printf("O(n²) demo with N=%d took %.3f ms\n", N, float64(t1.Sub(t0).Microseconds())/1000)
}
